test/sync/src/commonMain/kotlin/io/realm/test/mongodb/TestApp.kt

```diff
@@ -37,32 +37,62 @@ const val TEST_APP_1 = "testapp1" // Id for the default test app
  * This class merges the classes `App` and `AdminApi` making it easier to create an App that can be
  * used for testing.
  *
- * @param logLevel log level used to prime the AppConfiguration.Builder.
- * @param builder the builder used to build the final app. The builder is already primed with the
- * default test app configuration, but can be used to override the defaults and add additional
- * options.
+ * @param app gives access to the [App] class delegate for testing purposes
  * @param debug enable trace of command server and rest api calls in the test app.
  */
 class TestApp(
-    appName: String = TEST_APP_1,
-    dispatcher: CoroutineDispatcher = singleThreadDispatcher("test-app-dispatcher"),
-    appId: String = runBlocking(dispatcher) { getAppId(appName, debug) },
-    logLevel: LogLevel = LogLevel.WARN,
-    builder: (AppConfiguration.Builder) -> AppConfiguration.Builder = { it },
+    val app: App,
     debug: Boolean = false
-) : App by App.create(builder(testAppConfigurationBuilder(appId, logLevel)).dispatcher(dispatcher).build()),
-    AdminApi by (runBlocking(dispatcher) { AdminApiImpl(TEST_SERVER_BASE_URL, appId, debug, dispatcher) }) {
+) : App by app,
+    AdminApi by (app.configuration.networkTransportDispatcher.let { dispatcher -> runBlocking(dispatcher) { AdminApiImpl(TEST_SERVER_BASE_URL, app.configuration.appId, debug, dispatcher) } }) {
+
+    /**
+     * Creates an [App] with the given configuration parameters.
+     *
+     * @param logLevel log level used to prime the AppConfiguration.Builder.
+     * @param builder the builder used to build the final app. The builder is already primed with the
+     * default test app configuration, but can be used to override the defaults and add additional
+     * options.
+     * @param debug enable trace of command server and rest api calls in the test app.
+     **/
+    constructor(
+        appName: String = TEST_APP_1,
+        dispatcher: CoroutineDispatcher = singleThreadDispatcher("test-app-dispatcher"),
+        appId: String = runBlocking(dispatcher) { getAppId(appName, debug) },
+        logLevel: LogLevel = LogLevel.WARN,
+        builder: (AppConfiguration.Builder) -> AppConfiguration.Builder = { it },
+        debug: Boolean = false
+    ) : this(
+        App.create(
+            builder(testAppConfigurationBuilder(appId, logLevel))
+                .dispatcher(dispatcher)
+                .build()
+        )
+    )
 
     fun close() {
+        // This is needed to "properly reset" all sessions across tests since deleting users
```

Comment:
Is this also needed if we wipe the app cache data #540? If so, where did the discussion on getting `reset` exposed in the C-API?

Reply:
The conversation is here: https://github.com/realm/realm-core/pull/5014
In short, Jonathan doesn't seem too keen about exposing it, nor is Yavor. However, Nikola seemed pretty certain exposing `reste_for_testing` is needed in case we are to make Realm metadata-related assertions in case we change app configurations across test runs. The status now is waiting for Jonathan to comment since he was the most adamant one against the proposed solution.

In any case, I discussed it with Christian and he suggested we could remove the files manually after each test, which is also nice especially after running macos tests. That's what #540 does.